cronback-services/src/api/auth_middleware.rs

```diff
@@ -200,3 +220,260 @@ pub async fn authenticate<B>(
 
     Ok(resp)
 }
+
+#[cfg(test)]
+mod tests {
+
+    use std::collections::HashSet;
+    use std::fmt::Debug;
+
+    use axum::routing::get;
+    use axum::{middleware, Router};
+    use cronback_api_model::admin::CreateAPIkeyRequest;
+    use hyper::{Body, StatusCode};
+    use tower::ServiceExt;
+
+    use super::*;
+    use crate::api::auth_store::AuthStore;
+    use crate::api::config::ApiSvcConfig;
+    use crate::api::ApiService;
+
+    async fn state() -> AuthenticationState {
+        let mut set = HashSet::new();
+        set.insert("adminkey1".to_string());
+        set.insert("adminkey2".to_string());
+
+        let config = ApiSvcConfig {
+            address: String::new(),
+            port: 123,
+            database_uri: String::new(),
+            admin_api_keys: set,
+            log_request_body: false,
+            log_response_body: false,
+        };
+
+        let db = ApiService::in_memory_database().await.unwrap();
+        let auth_store = AuthStore::new(db);
+        let authenticator = Authenticator::new(auth_store);
+
+        AuthenticationState {
+            authenticator,
+            config,
+        }
+    }
+
+    struct TestInput {
+        app: Router,
+        auth_header: Option<String>,
+        on_behalf_on_header: Option<String>,
+        expected_status: StatusCode,
+    }
+
+    impl Debug for TestInput {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            f.debug_struct("TestInput")
+                .field("auth_header", &self.auth_header)
+                .field("on_behalf_on_header", &self.on_behalf_on_header)
+                .field("expected_status", &self.expected_status)
+                .finish()
+        }
+    }
+
+    struct TestExpectations {
+        unauthenticated: StatusCode,
+        authenticated: StatusCode,
+        admin_no_project: StatusCode,
+        admin_with_project: StatusCode,
+        unknown_secret_key: StatusCode,
+    }
+
+    async fn run_tests(
```

Comment:
Should we put `#[track_caller]` here to get helpful messages when assertions fail? Not super sure if this works with `async` functions though.

Reply:
TIL. But it doesn't work with `track_caller` indeed: https://github.com/rust-lang/rust/issues/87417